Tweet search using Pagefind

Accudio opened the pull request:
This Pull Request adds [Pagefind](https://github.com/cloudcannon/pagefind)—a search widget with static index—for a simple text search of all tweets.

This implementation runs the Pagefind CLI on the output HTML on the `eleventy.after` event, which then generates the index and required JS and CSS for the widget within `/_pagefind/` in the output directory. At the moment it uses the default Pagefind widget placed on `/search`, not yet linked from the main page.

Things to check/consider:
- How do we want to link `/search` up from the index, or do we want to place the search there?
- It works and performs well on ~6,500 tweets, does it cope well with larger archives?
- Should we keep the default Pagefind widget or make a custom one?
- Pagefind supports [filtering](https://pagefind.app/docs/filtering/) and [sorting](https://pagefind.app/docs/sorts/), are these of interest?

nhoizey commented:
I confirm that it works very well, even with more than 100.000 tweets:

<img width="1152" alt="image" src="https://user-images.githubusercontent.com/78213/208270606-1f792436-88bf-4fa6-9cee-fcc87b3a7763.png">

It would be great to have this on the homepage, but:
- unfortunately it doesn't work when we search for an original twitter.com tweet URL
- I don't think having two search input would be a good idea for UX

@Accudio do you think there's a way to fix the search of twitter.com URLs?

Accudio commented:
@nhoizey Yep, we can add the original twitter URL as an indexable string using data attributes. I've made some additions:

- Added Tweet ID and the original Twitter URL to the index so the widget can replace the existing search;
- Index external link URLs also;
- Fixed `newest`, `popular` and any future meta pages from being indexed and showing duplicate results;
- Replaced the homepage search with Pagefind.

nhoizey commented:
@Accudio any idea why it finds "only" 14,424 results for `https://twitter.com/nhoizey/status/`, while I guess all 118,048 tweets should be found?

![image](https://user-images.githubusercontent.com/78213/208302652-120f17aa-e63f-4606-9efb-9862cc4f4a09.png)

It works on your archive:

![image](https://user-images.githubusercontent.com/78213/208302713-271f5fe4-0e95-480d-8db8-e1e93a5baf77.png)

nhoizey commented on index.11ty.js:
```diff
@@ -263,19 +263,15 @@ class Index extends Twitter {
 			<div class="tweets-search">
-				<div class="lo" style="--lo-margin-h: 1em; align-items: center;">
-					<div class="lo-c" style="flex-grow: 100">
-						<input type="url" id="tweet-url" required placeholder="Tweet URL" style="width: 100%">
```
@Accudio could we have a static search input, with a default form action using DuckDuckGo for example, that would be enhanced with PageFind, instead of a JS only search?

Accudio replied:
Sounds good, however one snag is that it would require the build to know where the archive is deployed. Currently we don't have any metadata for this—would we be happy to add a `metadata.url` property or similar?

Otherwise that would be fine, although we can't enhance a search input using the Pagefind default UI we could replace it once Pagefind has been initialised.

Having a solution properly progressively-enhancing would be possible but a bit more work as we'd have to replace the UI library. This might be worth doing anyway, as the accessibility and flexibility of the default one isn't ideal, but I probably wouldn't be able to work on it for a while.

nhoizey replied:
You're right, we would need the URL somewhere.

I guess we can ship it like that, and open an issue for future enhancements.

nhoizey replied:
- #51

Accudio commented:
> @Accudio any idea why it finds "only" 14,424 results for `https://twitter.com/nhoizey/status/`, while I guess all 118,048 tweets should be found?
> 
> _image removed_
> 
> It works on your archive:
> 
> _image removed_

How odd, no I'm not aware of why that might be. In theory every tweet page has the full twitter URL (as linked on the twitter icon) indexed. Can you get a higher result with some other query?

nhoizey commented:
> Can you get a higher result with some other query?

It looks like I can't.

Even searching for `nhoizey` gives less results:

![image](https://user-images.githubusercontent.com/78213/208410073-679b97df-ce78-4722-983e-b9c9cd2469cd.png)

Maybe it's only indexing a full tweet URL, not its parts.

It works for all full tweet URLs I tested, so it's ok for me.

nhoizey reviewed (approved):
Great enhancement to the archive, thanks @Accudio ! 👍
